<a href="https://colab.research.google.com/github/RishbhaJain/Google-Gemma-3n-Kaggle-challenge/blob/main/nb/Gemma3N_(4B)-Conversational.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Read our **[Gemma 3N Guide](https://docs.unsloth.ai/basics/gemma-3n-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [2]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1,<4.0.0" huggingface_hub hf_transfer
    !pip install --no-deps unsloth

In [3]:
%%capture
# Install latest transformers for Gemma 3N
!pip install --no-deps --upgrade timm # Only for Gemma 3N

### Unsloth

`FastModel` supports loading nearly any model now! This includes Vision and Text models!

In [4]:
from unsloth import FastModel
import torch

fourbit_models = [
    # 4bit dynamic quants for superior accuracy and low memory use
    "unsloth/gemma-3n-E4B-it-unsloth-bnb-4bit",
    "unsloth/gemma-3n-E2B-it-unsloth-bnb-4bit",
    # Pretrained models
    "unsloth/gemma-3n-E4B-unsloth-bnb-4bit",
    "unsloth/gemma-3n-E2B-unsloth-bnb-4bit",

    # Other Gemma 3 quants
    "unsloth/gemma-3-1b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-4b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-12b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-27b-it-unsloth-bnb-4bit",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastModel.from_pretrained(
    model_name = "unsloth/gemma-3n-E4B-it",
    dtype = None, # None for auto detection
    max_seq_length = 1024, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.8.1: Fast Gemma3N patching. Transformers: 4.54.0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Gemma3N does not support SDPA - switching to eager!


model.safetensors.index.json: 0.00B [00:00, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/3.72G [00:00<?, ?B/s]

model-00002-of-00003.safetensors:   0%|          | 0.00/4.99G [00:00<?, ?B/s]

model-00003-of-00003.safetensors:   0%|          | 0.00/1.15G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/210 [00:00<?, ?B/s]

processor_config.json:   0%|          | 0.00/98.0 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

preprocessor_config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/4.70M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/777 [00:00<?, ?B/s]

# Gemma 3N can process Text, Vision and Audio!

Let's first experience how Gemma 3N can handle multimodal inputs. We use Gemma 3N's recommended settings of `temperature = 1.0, top_p = 0.95, top_k = 64`

In [5]:
from transformers import TextStreamer
# Helper function for inference
def do_gemma_3n_inference(messages, max_new_tokens = 128):
    _ = model.generate(
        **tokenizer.apply_chat_template(
            messages,
            add_generation_prompt = True, # Must add for generation
            tokenize = True,
            return_dict = True,
            return_tensors = "pt",
        ).to("cuda"),
        max_new_tokens = max_new_tokens,
        temperature = 1.0, top_p = 0.95, top_k = 64,
        streamer = TextStreamer(tokenizer, skip_prompt = True),
    )

In [9]:
# Test conversations (list of message lists)
test_conversations = [
    [{"role": "user", "content": [{"type": "text", "text": "Mujhe batao ki aaj mausam kaisa hai?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Aap kaise ho?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Kya tum mujhe ek joke suna sakte ho?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Mujhe ek Hinglish recipe do aloo ke parathe ke liye."}]}],
    [{"role": "user", "content": [{"type": "text", "text": "What's your naam?"}]}],
]

# Run inference for each prompt
for i, conversation in enumerate(test_conversations, 1):
    print(f"\n🟢 Prompt {i}: {conversation[0]['content'][0]['text']}")
    do_gemma_3n_inference(conversation, max_new_tokens=256)


🟢 Prompt 1: Mujhe batao ki aaj mausam kaisa hai?
आज मौसम के बारे में जानकारी देने के लिए, मुझे यह जानना होगा कि आप किस स्थान के बारे में पूछ रहे हैं। कृपया अपना शहर या क्षेत्र बताएं ताकि मैं आपको सटीक मौसम की जानकारी दे सकूं।

उदाहरण के लिए, आप कह सकते हैं:

* "दिल्ली में आज मौसम कैसा है?"
* "मुंबई में आज मौसम कैसा है?"
* "लखनऊ में आज मौसम कैसा है?"

एक बार जब आप स्थान बता देंगे, तो मैं आपको तापमान, आर्द्रता, हवा की गति और अन्य मौसम संबंधी जानकारी दे पाऊंगा।

आप मौसम के बारे में जानकारी के लिए Google पर भी खोज कर सकते हैं। बस "मौसम" टाइप करें और अपना शहर या क्षेत्र दर्ज करें।
<end_of_turn>

🟢 Prompt 2: Aap kaise ho?
मैं ठीक हूँ, धन्यवाद! आप कैसे हैं? 

(I am fine, thank you! How are you?)

I am a large language model, so I don't experience feelings like humans do, but I am functioning well and ready to help.  Let me know what you need!
<end_of_turn>

🟢 Prompt 3: Kya tum mujhe ek joke suna sakte ho?
ज़रूर, मैं तुम्हें एक मज़ाक सुनाता हूँ:

शिक्षक: "क्या होता है जब एक बिल्ली कंप्यूटर पर

# Let's finetune Gemma 3N!

You can finetune the vision and text parts for now through selection - the audio part can also be finetuned - we're working to make it selectable as well!

We now add LoRA adapters so we only need to update a small amount of parameters!

In [10]:
model = FastModel.get_peft_model(
    model,
    finetune_vision_layers     = False, # Turn off for just text!
    finetune_language_layers   = True,  # Should leave on!
    finetune_attention_modules = True,  # Attention good for GRPO
    finetune_mlp_modules       = True,  # SHould leave on always!

    r = 8,           # Larger = higher accuracy, but might overfit
    lora_alpha = 8,  # Recommended alpha == r at least
    lora_dropout = 0,
    bias = "none",
    random_state = 3407,
)

Unsloth: Making `model.base_model.model.model.language_model` require gradients


<a name="Data"></a>
### Data Prep
We now use the `Gemma-3` format for conversation style finetunes. We use [Maxime Labonne's FineTome-100k](https://huggingface.co/datasets/mlabonne/FineTome-100k) dataset in ShareGPT style. Gemma-3 renders multi turn conversations like below:

```
<bos><start_of_turn>user
Hello!<end_of_turn>
<start_of_turn>model
Hey there!<end_of_turn>
```

We use our `get_chat_template` function to get the correct chat template. We support `zephyr, chatml, mistral, llama, alpaca, vicuna, vicuna_old, phi3, llama3, phi4, qwen2.5, gemma3` and more.

In [11]:
from unsloth.chat_templates import get_chat_template
tokenizer = get_chat_template(
    tokenizer,
    chat_template = "gemma-3",
)

We get the first 3000 rows of the dataset

In [12]:
from datasets import load_dataset

dataset = load_dataset("Abhishekcr448/Hinglish-Everyday-Conversations-1M", split="train")
dataset = dataset.shuffle(seed=42).select(range(25000))  # limit for speed


README.md: 0.00B [00:00, ?B/s]

hinglish_conversations.csv:   0%|          | 0.00/185M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/1001323 [00:00<?, ? examples/s]

In [13]:
def to_chat_format(example):
    return {
        "messages": [
            {"role": "user", "content": example["input"]},
            {"role": "assistant", "content": example["output"]}
        ]
    }

dataset = dataset.map(to_chat_format)

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

In [14]:
dataset = dataset.remove_columns(["input", "output"])


We now use `standardize_data_formats` to try converting datasets to the correct format for finetuning purposes!

In [15]:
from unsloth.chat_templates import standardize_data_formats
dataset = standardize_data_formats(dataset)

Let's see how row 100 looks like!

In [11]:
dataset[100]

{'messages': [{'content': 'Yeh anniversary special hai, kya tumne kisi fancy restaurant ka naam suna?',
   'role': 'user'},
  {'content': 'Haan, maine suna hai, unka food aur ambiance bahut romantic hai, wahan chalein?',
   'role': 'assistant'}]}

We now have to apply the chat template for `Gemma-3` onto the conversations, and save it to `text`. We remove the `<bos>` token using removeprefix(`'<bos>'`) since we're finetuning. The Processor will add this token before training and the model expects only one.

In [16]:
def formatting_prompts_func(examples):
    convos = examples["messages"]  # correctly access the list of convos
    texts = [tokenizer.apply_chat_template(
        convo,
        tokenize=False,
        add_generation_prompt=False
    ) for convo in convos]
    return { "text": texts }

dataset = dataset.map(formatting_prompts_func, batched = True)

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Let's see how the chat template did! Notice there is no `<bos>` token as the processor tokenizer will be adding one.

In [13]:
dataset[100]

{'messages': [{'content': 'Yeh anniversary special hai, kya tumne kisi fancy restaurant ka naam suna?',
   'role': 'user'},
  {'content': 'Haan, maine suna hai, unka food aur ambiance bahut romantic hai, wahan chalein?',
   'role': 'assistant'}],
 'text': '<bos><start_of_turn>user\nYeh anniversary special hai, kya tumne kisi fancy restaurant ka naam suna?<end_of_turn>\n<start_of_turn>model\nHaan, maine suna hai, unka food aur ambiance bahut romantic hai, wahan chalein?<end_of_turn>\n'}

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`.

In [17]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    eval_dataset = None, # Can set up evaluation!
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 4, # Use GA to mimic batch size!
        warmup_steps = 5,
        num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 60,
        learning_rate = 2e-4, # Reduce to 2e-5 for long training runs
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/25000 [00:00<?, ? examples/s]

We also use Unsloth's `train_on_completions` method to only train on the assistant outputs and ignore the loss on the user's inputs. This helps increase accuracy of finetunes!

In [18]:
from unsloth.chat_templates import train_on_responses_only
trainer = train_on_responses_only(
    trainer,
    instruction_part = "<start_of_turn>user\n",
    response_part = "<start_of_turn>model\n",
)

Map (num_proc=2):   0%|          | 0/25000 [00:00<?, ? examples/s]

Let's verify masking the instruction part is done! Let's

*   List item
*   List item

print the 100th row again.  Notice how the sample only has a single `<bos>` as expected!

In [19]:
tokenizer.decode(trainer.train_dataset[100]["input_ids"])

'<bos><bos><start_of_turn>user\nYeh anniversary special hai, kya tumne kisi fancy restaurant ka naam suna?<end_of_turn>\n<start_of_turn>model\nHaan, maine suna hai, unka food aur ambiance bahut romantic hai, wahan chalein?<end_of_turn>\n'

Now let's print the masked out example - you should see only the answer is present:

In [20]:
tokenizer.decode([tokenizer.pad_token_id if x == -100 else x for x in trainer.train_dataset[100]["labels"]]).replace(tokenizer.pad_token, " ")

'                         Haan, maine suna hai, unka food aur ambiance bahut romantic hai, wahan chalein?<end_of_turn>\n'

In [18]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
12.592 GB of memory reserved.


# Let's train the model!

To resume a training run, set `trainer.train(resume_from_checkpoint = True)`

In [21]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 25,000 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 1 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (1 x 4 x 1) = 4
 "-____-"     Trainable parameters = 19,210,240 of 7,869,188,432 (0.24% trained)
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.


Unsloth: Will smartly offload gradients to save VRAM!


In [20]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

554.5282 seconds used for training.
9.24 minutes used for training.
Peak reserved memory = 12.592 GB.
Peak reserved memory for training = 0.0 GB.
Peak reserved memory % of max memory = 85.422 %.
Peak reserved memory for training % of max memory = 0.0 %.


<a name="Inference"></a>
### Inference
Let's run the model via Unsloth native inference! According to the `Gemma-3` team, the recommended settings for inference are `temperature = 1.0, top_p = 0.95, top_k = 64`

In [23]:
from unsloth.chat_templates import get_chat_template
from transformers import TextStreamer

# Register Unsloth's chat template
tokenizer = get_chat_template(tokenizer, chat_template="gemma-3")

# Define test prompts in chat format
test_conversations = [
    [{"role": "user", "content": [{"type": "text", "text": "Mujhe batao ki aaj mausam kaisa hai?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Aap kaise ho?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Kya tum mujhe ek joke suna sakte ho?"}]}],
    [{"role": "user", "content": [{"type": "text", "text": "Mujhe ek Hinglish recipe do aloo ke parathe ke liye."}]}],
    [{"role": "user", "content": [{"type": "text", "text": "What's your naam?"}]}],
]

# Run generation for each conversation
for i, convo in enumerate(test_conversations, 1):
    print(f"\n🟢 Prompt {i}: {convo[0]['content'][0]['text']}")

    # Tokenize the current conversation
    inputs = tokenizer.apply_chat_template(
        convo,
        add_generation_prompt=True,
        tokenize=True,
        return_tensors="pt",
        return_dict=True,
    ).to("cuda")

    # Generate output
    outputs = model.generate(
        **inputs,
        max_new_tokens=64,
        temperature=1.0,
        top_p=0.95,
        top_k=64,
    )

    # Decode and print
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    print(response)



🟢 Prompt 1: Mujhe batao ki aaj mausam kaisa hai?
user
Mujhe batao ki aaj mausam kaisa hai?
model
Aaj mausam bahut achha hai, halki hawa aur samudra ka dekho!

🟢 Prompt 2: Aap kaise ho?
user
Aap kaise ho?
model
Main theek hoon, aap kaise ho?

🟢 Prompt 3: Kya tum mujhe ek joke suna sakte ho?
user
Kya tum mujhe ek joke suna sakte ho?
model
Bilkul! Kyun computer ko doctor ke paas nahi jaha?

🟢 Prompt 4: Mujhe ek Hinglish recipe do aloo ke parathe ke liye.
user
Mujhe ek Hinglish recipe do aloo ke parathe ke liye.
model
Okay, here's a Hinglish recipe for Aloo Ki Parathe, super easy and delicious!

**Aloo Ki Parathe - Hinglish Style**

**Ingredients:**

*   2-3 medium-sized potatoes (aloo), boiled and mashed
*   1/2 tsp hing (asafo

🟢 Prompt 5: What's your naam?
user
What's your naam?
model
Main hoon Bard, Google DeepMind dwara train kiya gaya ek bada bhasha model.


 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [25]:
model.save_pretrained("gemma-3n")  # Local saving
tokenizer.save_pretrained("gemma-3n")


HfHubHTTPError: (Request ID: Root=1-6890d8fd-4a800c7408ec393d2a589143;317af361-bd0e-4639-90c0-98016848c9b6)

403 Forbidden: You don't have the rights to create a model under the namespace "HF_ACCOUNT".
Cannot access content at: https://huggingface.co/api/repos/create.
Make sure your token has the correct permissions.

In [27]:
model.push_to_hub("RJ-1411/gemma-3", token="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")
tokenizer.push_to_hub("RJ-1411/gemma-3", token="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")

README.md:   0%|          | 0.00/604 [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

adapter_model.safetensors:   0%|          | 0.00/76.9M [00:00<?, ?B/s]

Saved model to https://huggingface.co/RJ-1411/gemma-3


  0%|          | 0/2 [00:00<?, ?it/s]

tokenizer.model:   0%|          | 0.00/4.70M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

In [ ]:
from unsloth.chat_templates import get_chat_template
tokenizer = get_chat_template(
    tokenizer,
    chat_template = "gemma-3",
)

inputs = tokenizer.apply_chat_template(
    test_messages,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
    tokenize = True,
    return_dict = True,
).to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens = 64, # Increase for longer outputs!
    # Recommended Gemma-3 settings!
    temperature = 1.0, top_p = 0.95, top_k = 64,
)
tokenizer.batch_decode(outputs)

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch
from peft import PeftModel

# Load Base (pretrained only) Model
base_model = AutoModelForCausalLM.from_pretrained("unsloth/gemma-3n-E4B-it")
tokenizer = AutoTokenizer.from_pretrained("unsloth/gemma-3n-E4B-it")


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

### Saving to float16 for VLLM

We also support saving to `float16` directly for deployment! We save it in the folder `gemma-3N-finetune`. Set `if False` to `if True` to let it run!

If you want to upload / push to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now for all models! For now, you can convert easily to `Q8_0, F16 or BF16` precision. `Q4_K_M` for 4bit will come later!

Likewise, if you want to instead push to GGUF to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

Now, use the `gemma-3N-finetune.gguf` file or `gemma-3N-finetune-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
